flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSource.java

```diff
@@ -107,48 +127,74 @@ public Boundedness getBoundedness() {
     }
 
     /**
-     * Converts checkpoint between sources to transfer end position to next source's start position.
-     * Only required for dynamic position transfer at time of switching, otherwise source can be
+     * Callback for switch time customization of the underlying source from previous enumerator end
+     * state.
+     *
+     * <p>Called when the current enumerator has finished and before the next enumerator is created.
+     * The enumerator end state can thus be used to set the next source's start start position.
+     *
+     * <p>Only required for dynamic position transfer at time of switching, otherwise source can be
      * preconfigured with a start position during job submission.
      */
-    public interface CheckpointConverter<InCheckpointT, OutCheckpointT>
-            extends Function<InCheckpointT, OutCheckpointT>, Serializable {}
+    public interface SourceConfigurer<SourceT extends Source, FromEnumT extends SplitEnumerator>
```

Comment:
Yes this approach seems more flexible but does it imply that we need to touch all sources to allow splits being set?

It feels like we end up with mutable sources then instead of setting the changes in their builders? Maybe that also implies that we should actually have builders in the `SourceListEntry` but then we need some common interfaces for builders (something that I actually wanted to do independently).

Reply:
We would either need to allow to set start position or splits on the source or clone the existing source and modify that. For now, this hook provides the flexibility to solve set start position in any way a source allows, in a source specific way. It is backward compatible (we need this to work with 1.12) and a customization can resort to reflection if need be. Future extensions can be built on top of it, like a function that recognizes optional interfaces. Preferably we don't widen the scope of this PR.

Will update javadoc to convey the intent.

Reply:
Hm, this means that every source needs to provide a setter to be usable in `HybridSource`. We probably can't even force that with an API change as start offsets are defined very differently.

I'm still thinking that the builder route would go more with the intended design of `Source` but we can also keep as is for now and see how it affects `FileSource` and `KafkaSource` first.

Reply:
If source mutability is the only concern, then alternatively the function can return a cloned source with the start position augmented. I'm still leaning slightly towards the former checkpoint based solution but there are pros and cons for each of the options that have been discussed so far. I took this back to the mailing list thread, let's continue here and try to settle on an approach soon!